src/joy_api.c

```diff
@@ -109,75 +109,137 @@ static unsigned int joy_splt_format_data(flow_record_t *rec,
                                          unsigned char *data)
 {
     unsigned int i = 0;
+    unsigned int entry_counter = 0;
     unsigned int entries_used = 0;
     unsigned int num_of_pkts = 0;
     unsigned int data_len = 0;
+    uint16_t last_len = 0;
+    int16_t run_count = 0;
     struct timeval ts;
     uint16_t *formatted_data = (uint16_t*)data;
 
-    /* see how many packets we have to process - max is MAX_NFV9_SPLT_SALT_PKTS */
-    num_of_pkts = (rec->op < MAX_NFV9_SPLT_SALT_PKTS) ? rec->op : MAX_NFV9_SPLT_SALT_PKTS;
+    /* figure out actual entries needed
+     * we max out at 10 entries for a SPLT record.
+     */
+    for (i=0; i < rec->op; ++i) {
+        if (rec->pkt_len[i] != last_len ) {
+            /* run has ended increment entry count */
+            if (run_count < 0) {
+                ++entry_counter;
+            }
+
+            /* store last length */
+            last_len = rec->pkt_len[i];
+
+            /* increment entry count, reset run */
+            ++entry_counter;
+            run_count = 0;
+
+            /* stop if we reach the entry limit */
+            if (entry_counter == MAX_NFV9_SPLT_SALT_PKTS) {
+                break;
+            }
+        } else {
+            /* found a run, start the count */
+            --run_count;
+
+            /* stop if we reach the entry limit */
+            if (entry_counter == (MAX_NFV9_SPLT_SALT_PKTS-1)) {
+                break;
+            }
+        }
+    }
+    /* if we had an active run, increment entry count
+     * this is why the run count leg break 1 entry early
+     */
+    if (run_count < 0) {
+        ++entry_counter;
+    }
+
+    /* number of packets needed to create entries */
+    num_of_pkts = ++i;
+
+    /* if there were runs, adjust the entries used count */
+    if (entry_counter < num_of_pkts) {
+        entries_used = entry_counter;
+    } else {
+        entries_used = num_of_pkts;
+    }
 
     /* figure out the length of the data we are formatting */
     if (export_frmt == JOY_NFV9_EXPORT) {
         /* NFv9 is always 40 bytes */
         data_len = MAX_NFV9_SPLT_SALT_ARRAY_LENGTH;
+        entries_used = MAX_NFV9_SPLT_SALT_PKTS;
     } else {
         /* IPFix is variable length - each entry is represented by 2 16-bit values */
-        data_len = num_of_pkts * 4;
-    }
-
-    /* format for Netflow v9 export */
-    if (export_frmt == JOY_NFV9_EXPORT) {
-        /* loop through the SPLT lengths and store appropriately */
-        for (i=0; i < num_of_pkts; ++i) {
-            *(formatted_data+i) = (uint16_t)rec->pkt_len[i];
-        }
-
-        if (num_of_pkts < MAX_NFV9_SPLT_SALT_PKTS) {
-            /* padding needs to occur */
-            for (;i < MAX_NFV9_SPLT_SALT_PKTS; ++i) {
-               *(formatted_data+i) = (uint16_t)-32768;
+        data_len = entries_used * 4;
```

Comment:
#define ENTRY_SIZE sizeof(uint16_t)*2 just suggestion :)

Reply:
4 actually reads better with the specification in this case.